src/dataflow/decode/csv.rs

```diff
@@ -42,32 +50,86 @@ where
+                                        let length = bounds.len();
+                                        bounds.extend(std::iter::repeat(0).take(length));
+                                    }
+                                    csv_core::ReadRecordResult::Record => {
+                                        if bounds_valid != n_cols {
+                                            events_error += 1;
+                                            error!(
+                                                "CSV error: expected {} columns, got {}. Ignoring row.",
+                                                n_cols, bounds_valid,
+                                            );
+                                        } else {
+                                            events_success += 1;
+                                            session.give((
+                                                Row::pack(
+                                                    (0..n_cols)
+                                                        .map(|i| {
+                                                            // Unsafety rationalized as 1. the input text is determined to be
+                                                            // valid utf8, and 2. the delimiter is ascii, which should make each
+                                                            // delimited region also utf8.
+                                                            Datum::String(unsafe {
+                                                                std::str::from_utf8_unchecked(
+                                                                    &buffer
+                                                                        [bounds[i]..bounds[i + 1]],
+                                                                )
+                                                            })
+                                                        })
+                                                        .chain(iter::once(
+                                                            line_no.map(Datum::Int64).into(),
+                                                        )),
+                                                ),
+                                                *cap.time(),
+                                                1,
+                                            ));
+                                            // Reset valid data to extract the next record, should one exist.
+                                            buffer_valid = 0;
+                                            bounds_valid = 0;
+                                        }
+                                    }
+                                    csv_core::ReadRecordResult::End => {
+                                        done = true;
+                                    }
+                                }
```

Comment:
The nesting here is pretty brutal. Don't let it stand in the way of this PR, but this function could definitely use some refactoring.